job-info: Store full job-history, allow users to query pending, running, and inactive jobs

chu11 opened the pull request:
This is an initial implementation for #2454.  This will not pass travis right now due to #2470.

- In this initial implementation, the internal lists of jobs in the `job-info` module grow without bound.

- Restart by loading job history from the KVS is not supported in this round.

- Per discussion, in #2454, three lists are used to store jobs.  `pending` jobs are sorted by priority first, then submission time.  `running` jobs are sorted by start time of running.  And `inactive` jobs are stored by time completed.  Note that the latter two lists don't need to be "sorted" per-se.  You can just append to the end of the list when events happen.

- Flags are supported to allow callers to ask for the `pending`, `running`, or `inactive` jobs.  `flux job list` defaults to `PENDING | RUNNING`.

- Per discussion, in order to maintain tests against `job-manager.list`, add new `t/job-manager/list-jobs` tool was added.

- Right now only instance owners can query.  That can be changed easily to `FLUX_ROLE_USER`, but wasn't sure if guest access needs to be supported?  Or filtering based on a userid needs to be supported?  Or would that be for a porcelain tool (#2416)?

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -0,0 +1,479 @@
+    zlistx_t *running;
+    zlistx_t *inactive;
+    zlistx_t *processing;
+    zlist_t *futures;
```
"One of these things is not like the others; one of these things does not belong..."
Ahem sorry, Sesame Street.  Should the lone zlist_t be a zlistx_t just to avoid having both API's mixed in the same module?

chu11 replied:
hehe, `zlistx_t` doesn't have some friendly functions like `zlist_pop()`.  Let me make it work with the `zlistx_t` interface.

chu11 replied:
all of 30 seconds into trying to adjust to `zlistx_t`, I remember why I kept it as `zlist_t`.  `zlistx_t` uses handles for detach/delete.  But since this list is for keeping track of futures and not my own personal data structure, I'd have to do `flux_future_aux_set()` to save off the handle in the future.  Not sure if it's a net win.  I'll code it up and see.

garlick replied:
`zlist_remove()` searches the list for the item, so the equivalent in zlistx would just be
```c
void *handle = zlistx_find (futures, f);
if (handle)
    zlistx_delete (futures, handle); // or zlistx_detach()
```

chu11 replied:
ahh good idea.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -0,0 +1,479 @@
+                             zlistx_t *oldlist,
+                             flux_job_state_t newstate)
+{
+    if (zlistx_delete (oldlist, job->list_handle) < 0)
```
even though there's no destructor set on this list, `zlistx_detach()` might be a better choice here for clarity.

garlick commented on src/modules/job-info/job_state.c:
```diff
@@ -0,0 +1,479 @@
+ * The list `futures` is used to store in process futures.
+ */
+
+struct job_state_ctx {
```
Also defined in job_state.h?

chu11 replied:
hmmm strange, maybe I pushed the wrong version of my tree.

chu11 commented:
> > Right now only instance owners can query.
> 
> That does not seem to be the case in the code? Which is good, all users should be able to query.

Oh whoops, I got confused on the rolemasks, rolemask of `0` means all can send the RPC.

garlick commented:
> Oh whoops, I got confused on the rolemasks, rolemask of 0 means all can send the RPC.

No, 0 means instance owner only.  But I thought in the code it says FLUX_ROLE_USER?   Did I misread?

chu11 commented:
> No, 0 means instance owner only. But I thought in the code it says FLUX_ROLE_USER? Did I misread?

Ahh, no you didn't.

>     { .typemask     = FLUX_MSGTYPE_REQUEST,
>      .topic_glob   = "job-info.list",
>      .cb           = list_cb,
>      .rolemask     = FLUX_ROLE_USER
>    },

I got confused on the `job-state` callback vs the `job-info.list` callback.

chu11 commented:
Re-pushed with adjustments for review comments above

- use `zlistx_t` for all lists
- support filter job listing based on userid as well as job group
- update tests to avoid issue noted in #2470
- add some TODO comments for tests to do later

Note that I have not added any tests to get jobs for a guest userid.  I wasn't sure how to make the tests non-racy.  Guest user submissions have to be faked by modifying the userid field in the job eventlog.  Outside of instrumentation, there's no way to stop the `job-info` module from reading that information after the `job-manager` plugin has advertised a job state transition.

chu11 commented:
hmmm, somehow this failed in one builder.  Not sure how it could be racy.

```
expecting success: 
	test $(${LIST_JOBS} | wc -l) -eq 1

not ok 11 - job-manager: queue contains 1 jobs
```

chu11 commented:
re-pushed, fixing minor racyness

chu11 commented:
along with

```
expecting success: 
	test $(${LIST_JOBS} | wc -l) -eq 1

not ok 11 - job-manager: queue contains 1 jobs
```

I've seen

```
./sharness.sh: line 339:  1395 Segmentation fault      (core dumped) ${LIST_JOBS} > list10.out
not ok 21 - job-manager: jobs are listed in submit order
```

a few times.  I'm guessing the tool can segfault somehow, but I'm at a loss how.  Hopefully can printf my way to victory.

chu11 commented:
> a few times. I'm guessing the tool can segfault somehow, but I'm at a loss how. Hopefully can printf my way to victory.

Ahh issue was, `json_pack` with a 'o' vs 'O'.

grondo commented:
> Ahh issue was, json_pack with a 'o' vs 'O'.

Nice find!

chu11 commented:
hmm, had one builder hit

```
ERROR: t2202-job-manager.t - exited with status 137 (terminated by signal 9?)
```

all tests passed though.  Hopefully travis was just slow.

chu11 commented:
re-pushed, added a few more tests for EPROTO / EINVAL on list protocol.

garlick commented:
This getting ready for a more detailed review?

chu11 commented:
Yup

chu11 commented on src/modules/job-info/job_state.c:
```diff
@@ -0,0 +1,482 @@
+        return jsctx->pending;
+    else if (state == FLUX_JOB_RUN
+             || state == FLUX_JOB_CLEANUP)
+        return jsctx->running;
```
I normally indent this way with else if :-)  Do you mean you prefer the ORs to line up?  Or am I not seeing something obvious.

```
    if (state == FLUX_JOB_DEPEND
             || state == FLUX_JOB_SCHED)
        return jsctx->pending;
    else if (state == FLUX_JOB_RUN
             || state == FLUX_JOB_CLEANUP)
        return jsctx->running;
```

garlick replied:
The ORs don't line up.  If that's the way you like it, I'm OK, just thought it was an oversight.

chu11 commented on src/common/libjob/job.c:
```diff
@@ -127,21 +127,32 @@ int flux_job_submit_get_id (flux_future_t *f, flux_jobid_t *jobid)
+                       | FLUX_JOB_LIST_INACTIVE);
     int saved_errno;
 
+    /* No flags supported yet */
```
ugh, i actually remember removing that.  Guess I screwed up a rebase.

chu11 commented on src/modules/job-info/list.c:
```diff
@@ -135,26 +138,32 @@ json_t *list_jobs (struct info_ctx *ctx, int max_entries, json_t *attrs)
+        if (!ret) {
+            if ((ret = list_job_array (jobs,
+                                       ctx->jsctx->running,
+                                       max_entries,
```
oh good catch, that's a huge oversight by me.  Should add some unit tests too.

chu11 replied:
Oh wait, things work b/c we are only returning only one jobs array to the user.  So the check below in `list_job_array()` does the trick.

```
            if (json_array_size (jobs) == max_entries)                                                                                                                                                             
                return 1;  
```

garlick replied:
OK, any way to make that behavior more evident in the function name or args?  (e.g. call it `list_job_append()` and pass `&jobs` with it initially set to NULL?   Since it's working no big deal if you leave as is.

chu11 replied:
renaming to `list_job_append` seems like a good idea

chu11 commented on src/modules/job-info/list.c:
```diff
@@ -0,0 +1,203 @@
+        }
+        if (!strcmp (attr, "id")) {
+            val = json_integer (job->id);
+        }
```
sounds good to me, do you think we should copy this idea into job-manager as well?

chu11 replied:
now that i think about it, probably isn't necessary to go into job-manager, as that service is more for testing rather than actual job listing.

garlick replied:
Yeah, in fact I'll probably want to pare it back once your stuff goes in.

chu11 commented:
>  An RPC to request the list of valid job attribute names would be useful (a listing tool that allows a custom output format to be constructed could have a --list-attrs option or simliar).

Sounds like a good idea.

> Perhaps we should have a hard cap of 1000 or so jobs that can be returned to prevent DoS from returning (potentially) millions of jobs in one response message

makes sense

>  We had discussed that a listing tool could return summary stats along with the job listings. Would it make sense to tack on a "summary" object to the list response that includes (say) the number of jobs in each state, for all users? (Just keep a running count as state transitions are received) Alternatively, a separate RPC.

Would putting this in the `stats` for job-info make sense?  Alternate RPC seems to make more sense if not in `stats`.

> It might be useful to subscribe to the exception event and track the highest priority (lowest) exception level received per job, then return that under an "exception" attribute. A listing tool could flag jobs that have received exceptions.

Sounds like a good idea. I think the first two points should go along in this PR, but I think the latter 2 would probably be better off in their own PR.

garlick commented:
> Sounds like a good idea. I think the first two points should go along in this PR, but I think the latter 2 would probably be better off in their own PR.

Works for me.

chu11 commented:
re-pushed & re-based, addressing review comments above.  Side notes:

- `list` service returns `id` of jobs all the time to users, additional attrs returned based on user inputs
- `list_job_array()` renamed to `list_job_append()`
- added a `list-attrs` service.  Which is a pretty trivial implementation, returning a hard coded array of attribute names.

garlick commented:
I think this is pretty close.

I'm thinking running and inactive jobs should be listed in reverse start/completion time order, so when we truncate the list, we see the most recently started/running jobs, not the oldest.  (Sorry I didn't realize that until I played around with the tool just now - easy to fix though I think).

I'd be OK with merging once that's done, and then we can look at follow on work such as a couple of suggestions from #2454 like plumbing tool outputs json, optional per state limits to list all running but a limited number of pending/inactive jobs like squeue.

chu11 commented:
> I'm thinking running and inactive jobs should be listed in reverse start/completion time order, so when we truncate the list, we see the most recently started/running jobs, not the oldest. (Sorry I didn't realize that until I played around with the tool just now - easy to fix though I think).

Ahh, now that you mention it, that should have been the obvious thing to do :-)

Pre-pending to the lists instead of append is easy, then have to tweak tests for expected output.

chu11 commented:
rebased & re-pushed with minor tweaks, if all looks good, just need to squash

grondo commented on src/cmd/flux-job.c:
```diff
@@ -69,6 +69,19 @@ static struct optparse_option list_opts[] =  {
+      .usage = "List inactive jobs",
+    },
+    { .name = "userid", .key = 'u', .has_arg = 1,
+      .usage = "Limit output to specific userid. " \
```
Minor: might want to add a `.arginfo = "UID"` or something here so that `--userid` shows as taking an option argument in `--help` output.

chu11 replied:
good idea

chu11 commented:
re-pushed with two tiny fixes per @grondo comments above.  Went ahead and squashed everything.

garlick commented:
I'm fine if you want to set merge-when-passing.
